### ISSUE TITLE
question - nil value on the dictionary

### ISSUE THREAD

**Issue by lucassilvagc** [priore/SOAPEngine#197]
Hi,

I'm using SOAPEngine for research purposes on web services and I'm trying to fetch data from my web service running on another machine. The fetching works just fine, everything seems to load accordingly:
```
[AnyHashable("Body"): Optional({
    consultaDadosResponse =     {
        consultaDadosReturn =         {
            nome = Marcos;
            saldo = "900.0";
        };
    };
})]
```
The problem is that, when I try to pass these values from the dictionary to a let I've just created (`let dadosRec:Dictionary = dict as Dictionary, then self.dados = dadosRec["..."] as! NSArray`), it actually returns null values, so nothing is parsed at all. The part between quotes should be where my results are, but the value doesn't matter. Except for Body, any tries will result in _nil_ being returned.

What should I do on this specific case? Is there something I'm missing?

Oh, for the record: this is based on the Swift example repository and the example on README.md.

Thanks in advance for the help, this plugin is amazing. 

BR.

**Comment by priore** [comment]
Hello,
`consultaDadosReturn` key is a Dicitonary (and you can use `let dadosRet = dict as Dictionary`) but it contains two keys (nome and saldo, and you can user` let nome = dadosRet["nome"] as String` or `let nome = dadosRet["saldo"] as String`), not contains a NSArray.

**Comment by priore** [comment]
if you think it can only contain one name, then it as an dictionary, or return more names, then as an array, you need to cast based on the type, like as follows:

```swift
if let rec = dict as? Dictionary {
   // single record
} else if let recs = dict as? NSArray {
  // multi record
}
```

**Comment by lucassilvagc** [comment]
Hi, thanks for the quick reply.

I tried your solutions and couldn't go too far. 
When using `let dadosRet = dict as Dictionary` and `let nome = dadosRet["nome"] as String`, I do need to "force" the conversion to String, otherwise it won't work at all. When I force it, this happens: 

> Thread 1: Fatal error: Unexpectedly found nil while unwrapping an Optional value

Right now, I'm stuck into receiving the data. All other parts of my application (everything that I need to send over to the WS) works perfectly, my only problem being receiving data. 

BTW, this is my current code:
`let soap = SOAPEngine()
        soap.delegate = self
        soap.actionNamespaceSlash = true
        soap.responseHeader = true
        let idCampo = Int(campoId.text!)
        soap.setIntegerValue(idCampo!, forKey: "idConta")
        soap.requestURL("http://192.168.0.10:8080/WSBanco/services/ServicoConta", soapAction: "consultarDadosContaiOS", completeWithDictionary: { (statusCode : Int, dict : [AnyHashable : Any]?) -> Void in`
And also, the output of dadosRec stays mostly the same.

Again, thanks for your help. Best regards!

**Comment by lucassilvagc** [comment]
If I select the Body on dadosRec, this is my output:

```
Optional({
    consultarDadosContaiOSResponse =     {
        consultarDadosContaiOSReturn =         {
            ativo = true;
            idconta = 1;
            nome = Marcos;
            saldo = "93000.0";
        };
    };
})
```

**Comment by priore** [comment]
Hello,
`actionNamespaceSlash` is need only if you see "_SOAPEngine no valid response, try again with actionNamespaceSlash_" message on debug console.

Since you are using `responseHeader`, you receive the complete soap message, it is a necessary feature for not full compatible soap services  with MS-SOAP standard, which means that must also use the response and return keys (`consultarDadosContaiOSResponse`, `consultDadosContaiOSReturn`) to be able to get the data.
```swift
let response = dict?["consultarDadosContaiOSResponse"] as? [AnyHashable: Any]
guard let ret = response?["consultDadosContaiOSReturn"] as? [AnyHashable: Any] else {
   // NO DATA
   return
 }

let ativo = ret["ativo"]
```

**Comment by lucassilvagc** [comment]
Hi, thanks again for your reply.

I just managed to make it work. As you pointed, since responseHeader generated Response and Return, I couldn't receive the data the way I presented originally. Trying to use your solution also presented no results, even after printing several times.

After I disabled responseHeader, I was able to parse the data and assign them to a var.

Thank you very much for your help and congratulations on your work!
